elk/training/eigen_reporter.py

```diff
@@ -0,0 +1,148 @@
+
+        self.register_buffer(
+            "contrastive_M2",
+            torch.zeros(in_features, in_features, device=device, dtype=dtype),
+        )
+        self.register_buffer(
+            "M2", torch.zeros(in_features, in_features, device=device, dtype=dtype)
+        )
+        self.register_buffer(
+            "invariance",
+            torch.zeros(in_features, in_features, device=device, dtype=dtype),
+        )
+
+    def forward(self, x: Tensor) -> Tensor:
+        """Return the raw score output of the probe on `x`."""
+        return self.linear(x).squeeze(-1)
+
+    def predict(self, x_pos: Tensor, x_neg: Tensor) -> Tensor:
+        return 0.5 * (self(x_pos) - self(x_neg))
+
+    @property
+    def neg_covariance(self) -> Tensor:
+        return self.contrastive_M2 / self.n
+
+    @property
+    def variance(self) -> Tensor:
+        return self.M2 / self.n
+
+    @torch.no_grad()
+    def update(self, x_pos: Tensor, x_neg: Tensor) -> None:
+        # Sanity checks
+        assert x_pos.ndim == 3, "x_pos must be of shape [batch, num_variants, d]"
+        assert x_pos.shape == x_neg.shape, "x_pos and x_neg must have the same shape"
+
+        # Average across variants inside each cluster, computing the centroids.
+        pos_centroids, neg_centroids = x_pos.mean(1), x_neg.mean(1)
+
+        # We don't actually call super because we need access to the earlier estimate
+        # of the population mean in order to update (cross-)covariances properly
+        # super().update(x_pos, x_neg)
+
+        sample_n = pos_centroids.shape[0]
+        self.n += sample_n
+
+        # Update the running means; super().update() does this usually
+        neg_delta = neg_centroids - self.neg_mean
+        pos_delta = pos_centroids - self.pos_mean
+        self.neg_mean += neg_delta.sum(dim=0) / self.n
+        self.pos_mean += pos_delta.sum(dim=0) / self.n
+
+        # *** Variance (inter-cluster) ***
+        # See code at https://bit.ly/3YC9BhH, as well as "Welford's online algorithm"
+        # in https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance.
+        # Post-mean update deltas are used to update the (co)variance
+        neg_delta2 = neg_centroids - self.neg_mean  # [n, d]
+        pos_delta2 = pos_centroids - self.pos_mean  # [n, d]
+        self.M2.addmm_(neg_delta.mT, neg_delta2)
+        self.M2.addmm_(pos_delta.mT, pos_delta2)
+
+        # *** Invariance (intra-cluster) ***
+        # This is just a standard online *mean* update, since we're computing the
+        # mean of covariance matrices, not the covariance matrix of means.
+        sample_invar = cov_mean_fused(x_pos) + cov_mean_fused(x_neg)
+        self.invariance += (sample_n / self.n) * (sample_invar - self.invariance)
+
+        # *** Negative covariance ***
+        self.contrastive_M2.addmm_(neg_delta.mT, pos_delta2)
+        self.contrastive_M2.addmm_(pos_delta.mT, neg_delta2)
+
+    def fit_streaming(self) -> float:
```

Comment:
Could the fit_streaming warm-start with the previous eigenvector and run power iteration until it re-converges? This might be faster

Reply:
currently doing this with the SciPy solver (CPU tensors)